docs/features/sinks/azure-application-insights.md

```diff
@@ -57,4 +57,66 @@ if(string.IsNullOrEmpty(key) == false)
 ILogger logger = loggerConfig.CreateLogger();
 ```
 
+### Q: Where can I initialize the logger in an ASP.NET Core application or other hosted service?
+
+The Application Insights instrumentation key is typically available via the configuration or via an `ISecreteProvider`.  It's best if the Logger is created and assigned to the Serilog `Log.Logger` as soon as possible.  The easiest place to do that in an ASP.NET application  is in the `ConfigureServices` method of the `Program` class.
+
+Another place where the logger can be initialized is in the `ConfigureAppConfiguration` method of the `IHostBuilder`:
+
+```csharp
+var host = Host.CreateDefaultBuilder()
+               .ConfigureAppConfiguration((context, configBuilder) =>
+               {
+                   var config = configBuilder.Build();
+
+                   var instrumentationKey = config["ApplicationInsights:InstrumentationKey"];
+
+                   var logConfiguration = new LoggerConfiguration()
+                                           .Enrich.FromLogContext()
+                                           .WriteTo.Console();
+
+                    var appInsightsInstrumentationKey = builtConfig["ApplicationInsights:InstrumentationKey"];
+
+                    if (!String.IsNullOrWhiteSpace(appInsightsInstrumentationKey))
+                    {
+                        logConfiguration.WriteTo.AzureApplicationInsights(appInsightsInstrumentationKey);
+                    }
+
+                    Log.Logger = logConfiguration.CreateLogger();
+               })
+               .UseSerilog()
+               .Build();
+```
+
+Or, if the Application Insights' instrumentation key is stored in a secretstore and you're making use of Arcus.Security:
+
+```csharp
+var host = Host.CreateDefaultBuilder()
+               .ConfigureSecretStore((context, config, builder) =>
+               {
+                   // Configure the secretstore here
+               })
+               .ConfigureServices((context, services) =>
+               {
+                   var serviceProvider = services.BuildServiceProvider();
```

Comment:
Building a service provider yourself is not recommended and is probably not a good idea to use as an example to newcomers.
If you need the service provider, look at this example in our templates: https://github.com/arcus-azure/arcus.templates/blob/master/src/Arcus.Templates.WebApi/Startup.cs#L306

And see the recommendation here: https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-3.1#recommendations

Reply:
Thanks for the link, interesting!

So, what do you suggest then ?  That we configure the Logger in the `ConfigureWebHost` method for ASP.NET applications ?
And for other 'hosted' applications, that we do this just after the host has been build, like this ?

```csharp
var host = Host.CreateDefaultBuilder(args)
                       .ConfigureServices( ... )
                       .ConfigureSecretStore ( .... )  
                       .UseConsoleLifetime()
                      .UseSeriLog()
                       .Build();

var secretProvider = host.Services.GetService<ISecretProvider>();

var instrumentationKey = secretProvider.GetRawSecretAsync("ApplicationInsights:InstrumentationKey").GetAwaiter().GetResult();
var logConfiguration = new LoggerConfiguration()
                              .Enrich.FromLogContext()
                              .WriteTo.Console();
if (!String.IsNullOrWhiteSpace(appInsightsInstrumentationKey))
{
    logConfiguration.WriteTo.AzureApplicationInsights(appInsightsInstrumentationKey);
}

Log.Logger = logConfiguration.CreateLogger();

 await host.RunAsync();
```

Reply:
Hmm, I think in the Arcus templates is a good place to look. We configure there actually the logger twice; once before the actual creation of the `IHost` so we definitely have a logger if the setup fails and we want to log some critical information; and while the the host is created so we can make use of the `IServiceProvider`. 
But I think the way it's set up; the instrumation key is mostly picked from the environment variables, via the configuration. The secret store is not something we have/could use easily here I think (like Tom said below).

Reply:
I've changed this so that the logger is configured in the UseSerilog extension method.
I've only added an example where the instrumentation key is retrieved from the configuration.   However, the UseSerilog extension method exposes an IServiceProvider as well, so I think we'll be able to retrieve the key from keyvault there as well.   Will need to try it first before I put it in the docs.